### PR TITLE
update.sh code refactoring

### PR DESCRIPTION
**Pull request by DL6ER**
**By submitting this pull request, I confirm the following (please check boxes, eg [X]) _Failure to fill the template will close your PR_:**

***Please submit all pull requests against the `development` branch. Failure to do so will delay or deny your request***

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] I have checked that [another pull request](https://github.com/pi-hole/pi-hole/pulls) for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.

**How familiar are you with the codebase?:** 10

---

Code refactoring of the existing `update.sh` without changing its external behavior. Internally, the behavior changes (getting completely independent of GitHub tags).

The old logic has been:

1. Get tag from GitHub
2. Get local tag
3. Compare tags and update the corresponding parts, if necessary

Proposed new logic (we discussed about that):

1. Fetch (possible) changes from `origin`
2. Get local git status
3. Update if local copy is behind remote version

Note that the new behavior is independent of tags but relies on the git log, only. Hence, it is much more reliable and trustworthy.

_This template was created based on the work of [`udemy-dl`](https://github.com/nishad/udemy-dl/blob/master/LICENSE)._

### CODE REVIEW

**Comment by DL6ER** [comment]
Note that this will also make the updater work for other branches, like:
```
root@raspberrypi:/etc/.pihole# git pull
Already up-to-date.
root@raspberrypi:/etc/.pihole# git reset --hard HEAD^
HEAD is now at 4632b0f Updated updater logic
root@raspberrypi:/etc/.pihole# pihole -up
::: Checking for updates...
::: Pi-hole Core:   update available
::: Web Interface:  up to date
:::
::: Pi-hole core files out of date
:::
[... rest not shown ...]
```

File: advanced/Scripts/update.sh [hunk 1]
```diff
@@ -40,22 +44,19 @@ make_repo() {
...
   # Stash all files not tracked for later retrieval
-  git -C "${directory}" stash --all --quiet &> /dev/null || ${retVal}=1
+  git -C "${directory}" stash --all --quiet
   # Force a clean working directory for cloning
-  git -C "${directory}" clean --force -d &> /dev/null || ${retVal}=1
+  git -C "${directory}" clean --force -d
   # Fetch latest changes and apply
-  git -C "${directory}" pull --quiet &> /dev/null || ${retVal}=1
-  return ${retVal}
+  git -C "${directory}" pull --quiet
```
Comment (Mcat12) [inline advanced/Scripts/update.sh]: We should use the long version of `-C` #1009

**Comment by dschaper** [comment]
`pihole -v` would show the tag for the release, but not the hash (currently.) Ideally, the update script shouldn't output any kind of information, just a return value that the caller would identify and then let the caller print the information. `update.sh` shouldn't ever be run bare, it should be called as a utility script/library function. Will the user ever see any of the output from this script?

**Comment by DL6ER** [comment]
It is run bare for updating.

https://github.com/pi-hole/pi-hole/blob/development/pihole#L51

**Comment by dschaper** [comment]
Right, it's called via a script though, we don't advise users to go to a shell prompt and run `update.sh` by itself.

**Comment by DL6ER** [comment]
Does that matter if is called via a script? Would you like to rewrite more fundamental?

**Comment by dschaper** [comment]
I'm over on MM now, but I think these should be utilities or the beginnings of a library and we handle all the input/output from `pihole`. So we keep the interface with the user in one place, instead of spreading it out. Longer explanation in chat...

**Comment by dschaper** [comment]
Approved

**Comment by dschaper** [comment]
Approved

**Pull request merged**